sensehat/src/main/java/com/google/android/things/contrib/driver/sensehat/Joystick.java

```diff
@@ -0,0 +1,170 @@
+@SuppressWarnings({"unused", "WeakerAccess"})
+public class Joystick implements AutoCloseable {
+
+    private static final String TAG = "SenseHatJoystick";
+
+    public static final int KEY_RELEASED = 0;
+    public static final int KEY_PRESSED_DOWN = 1;
+    public static final int KEY_PRESSED_RIGHT = 2;
+    public static final int KEY_PRESSED_UP = 4;
+    public static final int KEY_PRESSED_ENTER = 8;
+    public static final int KEY_PRESSED_LEFT = 16;
+
+    private static final int JOYSTICK_REG_KEYS = 0xF2;
+
+    private PeripheralManagerService mPioService;
+
+    private String mI2cBus;
+    private int mI2cAddress;
+    private Gpio mInterruptGpio;
+    private OnButtonEventListener mListener;
+    private int mPrevKey;
+
+    /**
+     * Interface definition for a callback to be invoked when a button event occurs.
+     */
+    public interface OnButtonEventListener {
+        /**
+         * Called when a button event occurs.
+         *
+         * @param key     the <code>KEY_PRESSED_*</code> key code of the button for which the event
+         *                occurred
+         * @param pressed <code>true</code> if the button is now pressed
+         */
+        void onButtonEvent(int key, boolean pressed);
+    }
+
+    /**
+     * Creates a new joystick driver.
+     *
+     * @param bus          the I2C bus the joystick controller is connected to
+     * @param address      the joystick controller I2C device slave address
+     * @param interruptPin the interrupt GPIO pin the joystick controller is connected to
+     * @throws IOException
+     */
+    public Joystick(String bus, int address, String interruptPin) throws IOException {
+        mPioService = new PeripheralManagerService();
+        Gpio interruptGpio = mPioService.openGpio(interruptPin);
+        try {
+            connect(bus, address, interruptGpio);
+        } catch (IOException | RuntimeException e) {
+            close();
+            throw e;
+        }
+    }
+
+    /**
+     * Constructor invoked from unit tests.
+     */
+    @VisibleForTesting
+    /*package*/ Joystick(String bus, int address, Gpio interruptGpio) throws IOException {
+        connect(bus, address, interruptGpio);
+    }
+
+    private void connect(String bus, int address, Gpio interruptGpio) throws IOException {
+        mI2cBus = bus;
+        mI2cAddress = address;
+        mInterruptGpio = interruptGpio;
+        mInterruptGpio.setDirection(Gpio.DIRECTION_IN);
+        mInterruptGpio.setEdgeTriggerType(Gpio.EDGE_BOTH);
+        mInterruptGpio.registerGpioCallback(mInterruptCallback);
+    }
+
+    /**
+     * Local callback to monitor GPIO edge events.
+     */
+    private GpioCallback mInterruptCallback = new GpioCallback() {
+        @Override
+        public boolean onGpioEdge(Gpio gpio) {
+            try {
+                boolean trigger = gpio.getValue();
+                if (trigger) {
+                    I2cDevice i2c = mPioService.openI2cDevice(mI2cBus, mI2cAddress);
```

Comment:
@proppy I suppose that this is indeed not a bug but the way I2C works (one active "opened" device at a time). Since all peripherals on Sense HAT shares the same I2C bus, the joystick works like this:

1. The ATtiny onboard receives joystick input.
2. ATtiny drives GPIO ```BCM23``` high as an interrupt signal for joystick event.
3. Raspberry Pi monitors ```BCM23```, and initiates I2C connection with ATtiny at ```0x46```.
4. Raspberry Pi reads the joystick key state via the I2C register at ```0xF2```.
5. Closes I2C connection afterwards to free up the bus for other peripherals.

So, the current commit should behave well with this protocol...